basic-api.yml

```diff
@@ -0,0 +1,142 @@
+      "lprNumber": "999-999-999",
+      "commuterClassification": "C1",
+      "birthCountry": "Bahamas",
+      "birthDate": "1958-07-17"
+    }
+  }],
+  "proof": [{
+    "type": "Ed25519Signature2018",
+    "created": "2020-02-25T22:03:05Z",
+    "jws": "eyJhbGciOiJFZERTQSIsImI2..hS5SPmWTFocjcDg",
+    "proofPurpose":"assertionMethod",
+    "verificationMethod":"did:key:z6Mkm8aUytv...nvKQrHB"
+  }]
+}
+        '400':
+          description: Malformed request from client.
+components:
+  schemas:
+    IssueRequest:
+      type: object
+      properties:
+        credential:
+          type: object
+          description: The credential to be issued minus the proofs
+          example: {
+  "@context": [
+    "https://www.w3.org/2018/credentials/v1",
+    "https://w3id.org/citizenship/v1"
+  ],
+  "id": "https://issuer.oidp.uscis.gov/credentials/83627465",
+  "type": [
+    "VerifiableCredential",
+    "PermanentResidentCard"
+  ],
+  "issuanceDate": "2019-12-03T12:19:52Z",
+  "expirationDate": "2029-12-03T12:19:52Z",
+  "credentialSubject": {
+    "id": "did:example:b34ca6cd37bbf23",
+    "type": [
+      "PermanentResident",
+      "Person"
+    ],
+    "givenName": "JOHN",
+    "familyName": "SMITH",
+    "gender": "Male",
+    "image": "data:image/png;base64,iVBORw0KGgo...kJggg==",
+    "residentSince": "2015-01-01",
+    "lprCategory": "C09",
+    "lprNumber": "999-999-999",
+    "commuterClassification": "C1",
+    "birthCountry": "Bahamas",
+    "birthDate": "1958-07-17"
+  }
+}
+        options:
+          type: object
+          description: Optional values that check or modify the input and modify the output of the request.
+          properties:
+            validation:
+              type: object
+              description: Optional validation schemas to use on the input document.
+              properties:
+                type:
+                  type: string
+                  description: The type of validation mechanism
+                  example: JsonSchema2018
+                credentialSchema:
+                  type: string
+                  description: A link to a credential schema to use for validation.
+                  example: https://example.com/schemas/prc.json
+            issuer:
+              type: string
+              description: Optional identifier for the issuer of the credential. If this is missing, it may be pre-configured or implied by other values in the request.
+              example: did:example:ec712ebc6f1c221ebfeb1f76e12
+            output:
+              type: object
+              description: Optional values that modify the output of the request.
+              properties:
+                representation:
+                  description: Optional value to specify the credential representation format. Well-known values are "jsonld" and "json". Other values MAY be supported. If this option is not specified a default of "jsonld" MUST be assumed.
+                  example: jsonld
```

Comment:
How does this work with the caller already passing in a fully formed VC (minus proof) in their representation of choice? Or is the input required to be jsonld (as in the example)?

Reply:
> How does this work with the caller already passing in a fully formed VC (minus proof) in their representation of choice? 

Excellent question, I think there are corner cases here that the group will need to discuss. In general, the design assumes that you pass in a fully formed VC (minus proof) in the representation format you expect to get out. So, JSON in - JSON out, or JSON-LD in - JSON-LD out.

There are future scenarios here where you might do JSON in, JSON-LD out, or JSON-LD in, CBOR out... or JSON in - JSON+JWT out (translation from one representation to another). 

We'll need to define what MUST be supported (where everything else is optional)

> Or is the request object required to be jsonld (as in the example)?

The way this is currently written, the request object is JSON (not JSON-LD). The credential is JSON-LD (which is always valid JSON)... everything else, like `options` is pure JSON. Or am I missing something?

Reply:
Oh, the "representation" is what you want to get back... not what you're passing in... maybe that's the confusion?

Reply:
The existing spec takes as input discrete pieces of the credential, while also providing an option to tune the output format. The service is in charge of producing a well-formed VC in the chosen format. This change proposes to change the input to a fully formed VC (minus proof) in the caller's chosen format while keeping the option to change the output format *if* the caller chooses to. The service's ability to produce correctly formatted VCs is now less valuable. 

What prompted the change from discrete parameters to a fully formed VC?

Reply:
@llorllale,

> The service's ability to produce correctly formatted VCs is now less valuable.

It would be nice for the service to do this. It seems that the thinking here is that the caller may have difficulty writing code that produces a valid VC and that the API will handle this for them. Of course, this means that the caller must produce a valid payload to send to the API -- which they can only do by either following yet another specification (the API specification) or by iteratively checking for errors against the API.

We already have a spec that says how to format at VC and it's a standard. The API caller just needs to comply with that specification. If we, instead, break up the VC into N different parts, now the API caller needs to use this *new* spec and correctly format N different things according to this entirely new specification. This is extra work and it perhaps does not achieve the goal I think you were describing. The API should return errors for malformed data either way. Furthermore, by directly surfacing all of the various *known* or *common* components of a VC at the API level, we create a problem whereby the API specification may need to change in order to support new features. There may be some middle ground we could find there -- but it comes at a complexity cost that may not actually be providing any significant benefit.

Reply:
> What prompted the change from discrete parameters to a fully formed VC?

The above reasoning from @OR13 and @dlongley, who I hadn't had the chance to sync up with yet, and they seem to have come to the same conclusions I did on the API, namely:

1. Let's focus on the simplest API that we can standardize on.
2. There are already specifications that tell you how to construct a valid VC, let's depend on those, as much as possible, rather than write new composition rules in this spec.

The more features we put into this API the more discussion we'll have on features that are not necessary for interop. That is not to say that there isn't value in providing a deconstructed VC to an endpoint, there probably is... but standardizing on that now feels like standardizing on a nice to have rather than standardizing on a necessity.

At some point all of us need to digitally sign a VC (allowing for a number of options/variations in the industry, like proof mechanism). Let's please focus on that instead of focusing on "nice to have" features. Clearly, if this API is too limited and doesn't address generating JWT-protected VCs (for example, I accidentally left out specifying the `kid`), then we do need to fix those things.

Reply:
There is a work in progress spec for sending discrete pieces to a verifier which may in turn construct and issue a credential... https://github.com/decentralized-identity/proof-presentation

^ this spec defines a json interface for providing information... which we could use here... but its clearly much more complex than just having the client do most of the work.

I'd rather take an iterative approach, get consensus on the simple case first, then tackle the harder ones, while we have something we can rely on for interop.

Reply:
Thank you for providing the reasoning - it's starting to make sense.